bigquery/params.go

```diff
@@ -127,14 +127,179 @@ type QueryParameter struct {
 	// above.  Null strings will report in query statistics as a valid empty
 	// string.
 	Value interface{}
+
+	explicitParameter Parameter
 }
 
-func (p QueryParameter) toBQ() (*bq.QueryParameter, error) {
+// ParameterType represents an explicit typed parameter to a query.
+// When using those interfaces, the priority is to avoid type inference.
+type ParameterType interface {
+	parameterType() *bq.QueryParameterType
+}
+
+// ParameterValue represents an explicit parameter value to a query
+// When using those interfaces, the priority is to avoid type inference.
+type ParameterValue interface {
+	parameterValue() *bq.QueryParameterValue
+}
+
+// Parameter implements ParameterType and ParameterValue.
+type Parameter interface {
+	ParameterType
+	ParameterValue
+}
+
+// ScalarQueryParameter represents an explicit scalar parameter with a given type.
+type ScalarQueryParameter struct {
+	// Name is used for named parameter mode.
+	// It must match the name in the query case-insensitively.
+	Name string
+
+	// Value is the value of the parameter.
+	//
+	// Same rules applies from QueryParameter.Value field.
+	// But as for explicit parameters we try to avoid type inference,
+	// we recommend setting the value as string using our formatter functions when applicable.
+	//   CivilTimeString
+	//   CivilDateTimeString
+	//   IntervalString
+	//   NumericString
+	//   BigNumericString
+	Value interface{}
+
+	// The parameter data type.
+	Type string
+}
+
+// ArrayQueryParameter represents an explicit array parameter with a given type.
+type ArrayQueryParameter struct {
+	// Name is used for named parameter mode.
+	// It must match the name in the query case-insensitively.
+	Name string
+
+	// Value is the value of the parameter.
+	//
+	// Same rules applies from QueryParameter.Value field.
+	// But as for explicit parameters we try to avoid type inference,
+	// we recommend setting the value as string using our formatter functions when applicable.
+	//   CivilTimeString
+	//   CivilDateTimeString
+	//   IntervalString
+	//   NumericString
+	//   BigNumericString
+	Value interface{}
+
+	// Type must conform to the ParameterType interface. You can have an array of scalar values or complex types.
+	Type ParameterType
+}
+
+// StructQueryParameter represents an explicit complex type parameter with a given schema.
+type StructQueryParameter struct {
+	// Name is used for named parameter mode.
+	// It must match the name in the query case-insensitively.
+	Name string
+
+	// Schema complex struct value passed as a map of field names and ParameterTypes.
+	Schema map[string]ParameterType
```

Comment:
leaf fields of a struct don't have to have names, so using a map will cause issues in those cases.

https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#struct_type

Reply:
I tried to add support for leaf fields without a name, but I could not find how to do that properly. I'm getting this error from the backend:

```
```

I was looking here on the API docs and they don't provide an example on how that can be done.
https://cloud.google.com/bigquery/docs/reference/rest/v2/QueryParameter#queryparametervalue

I also noticed that the Python API is based on the equivalent to a Go Map, that's why I started with that idea here. 
https://cloud.google.com/bigquery/docs/samples/bigquery-query-params-structs#bigquery_query_params_structs-python